src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -6659,6 +6660,193 @@ public static Task<Task<TResult>> WhenAny<TResult>(params Task<TResult>[] tasks)
+                }
+
+                return waiter;
+            }
+
+            /// <summary>Iterates through the tasks represented by the provided waiter.</summary>
+            public static async IAsyncEnumerable<T> Iterate<T>(WhenEachState? waiter, [EnumeratorCancellation] CancellationToken cancellationToken = default) where T : Task
+            {
+                // The enumerable could have GetAsyncEnumerator called on it multiple times. As we're dealing with Tasks that
+                // only ever transition from non-completed to completed, re-enumeration doesn't have much benefit, so we take
+                // advantage of the optimizations possible by not supporting that and simply have the semantics that, no matter
+                // how many times the enumerable is enumerated, every task is yielded only once. The original GetAsyncEnumerator
+                // call will give back all the tasks, and all subsequent iterations will be empty.
+                if (waiter?.TryStart() is true)
+                {
+                    // Loop until we've yielded all tasks.
+                    while (waiter.Remaining > 0)
+                    {
+                        // Either get the next completed task from the queue, or get a
+                        // ValueTask with which to wait for the next task to complete.
+                        Task? next;
+                        ValueTask waitTask = default;
+                        lock (waiter)
+                        {
+                            // Reset the MRVTSC if it was signaled, then try to dequeue a task and
+                            // either return one we got or return a ValueTask that will be signaled
+                            // when the next completed task is available.
+                            waiter._waitForNextCompletedTask.Reset();
+                            if (!waiter.TryDequeue(out next))
+                            {
+                                waitTask = new(waiter, waiter._waitForNextCompletedTask.Version);
+                            }
+                        }
+
+                        // If we got a completed Task, yield it.
+                        if (next is not null)
+                        {
+                            cancellationToken.ThrowIfCancellationRequested();
+                            waiter.Remaining--;
+                            yield return (T)next;
+                            continue;
+                        }
+
+                        // Otherwise, wait.
+                        if (cancellationToken.CanBeCanceled && !waitTask.IsCompleted)
+                        {
+                            // If we have a cancellation token and the ValueTask isn't already completed,
+                            // get a Task from the ValueTask so we can use WaitAsync to make the wait cancelable.
+                            await waitTask.AsTask().WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
I take it it's not worth the extra code to avoid the allocations here?

Reply:
I don't think it's worth it, at least not right now. If it became an issue, we could write more code to reduce some of the overhead, but there'd still be some.